test/functional/test_framework/test_framework.py

```diff
@@ -307,11 +307,12 @@ def shutdown(self):
                 node.cleanup_on_exit = False
             self.log.info("Note: bitcoinds were not stopped and may still be running")
 
+        exists_perf_subprocess = any(bool(node.perf_subprocesses) for node in self.nodes)
         should_clean_up = (
             not self.options.nocleanup and
             not self.options.noshutdown and
             self.success != TestStatus.FAILED and
-            not self.options.perf
+            (not self.options.perf or not exists_perf_subprocess)
```

Comment:
nit: alternative version, which I find easier to reason (feel free to ignore):
```suggestion
            not (self.options.perf and exists_perf_subprocess)
```

Reply:
Great